indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/ParallelIndexStatsReporter.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.common.task.batch.parallel;
+
+import com.google.common.collect.ImmutableMap;
+import org.apache.druid.indexing.common.IngestionStatsAndErrorsTaskReport;
+import org.apache.druid.indexing.common.IngestionStatsAndErrorsTaskReportData;
+import org.apache.druid.indexing.common.TaskReport;
+import org.apache.druid.java.util.common.ISE;
+import org.apache.druid.java.util.common.Pair;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.segment.incremental.ParseExceptionReport;
+import org.apache.druid.segment.incremental.RowIngestionMeters;
+import org.apache.druid.segment.incremental.RowIngestionMetersTotals;
+import org.apache.druid.segment.incremental.SimpleRowIngestionMeters;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public abstract class ParallelIndexStatsReporter
+{
+  private static final Logger LOG = new Logger(ParallelIndexStatsReporter.class);
+
+  // Row ingestion meters fields
+  private static final String PROCESSED_FIELD = "processed";
+  private static final String PROCESSED_BYTES_FIELD = "processedBytes";
+  private static final String PROCESSED_WITH_ERROR_FIELD = "processedWithError";
+  private static final String THROWN_AWAY_FIELD = "thrownAway";
+  private static final String UNPARSEABLE_FIELD = "unparseable";
+  // Ingestion stats and errors report fields
+  private static final String INGESTION_STATS_AND_ERRORS_FIELD = "ingestionStatsAndErrors";
+  private static final String PAYLOAD_FIELD = "payload";
+  private static final String ROW_STATS_FIELD = "rowStats";
+  private static final String TOTALS_FIELD = "totals";
+  private static final String UNPARSEABLE_EVENTS_FIELD = "unparseableEvents";
+
+  abstract ParallelIndexStats report(
```

Comment:
I took a look at this class and its sub-classes. I feel there is still some duplication between the single-phase and multi-phase implementations.

For the overall design, here is what I am thinking:
- We probably don't need multiple implementations and thus no need for the factory class either.
- Have a single implementation `ParallelIndexTaskStatsReporter`.
- Pass the task to its constructor.
- `report` method (maybe renamed `reportStats` or `getStats` for less ambiguity) should take the following
   - set of running task ids
   - map of reports (the new interace `TaskReportContainer` should be of use here)
   - we shouldn't have to pass the task again here, neither should we pass the indexingRunner as a plain `Object`.
- For parallel single-phase and multi-phase
  - the handling would be the same
- For sequential
   - we shouldn't be using `ParallelIndexTaskStatsReporter` at all because it is sequential, not parallel
   - we should continue to do what the original code was doing